products/data_modeling/backend/models/edge.py

```diff
@@ -1,18 +1,62 @@
-from django.db import models
+from django.db import connection, models
 
+from posthog.models import Team
 from posthog.models.utils import CreatedMetaFields, UpdatedMetaFields, UUIDModel
 
+from .node import Node
+
+
+class CycleDetectionError(Exception):
+    """The exception raised when an edge would cause a cycle in a DAG"""
+
+    pass
+
+
+class DAGMismatchError(Exception):
+    """exception raised when an edge would connect two different DAGs together"""
+
+    pass
+
+
+class DataModelingEdgeQuerySet(models.QuerySet):
+    def update(self, **kwargs):
+        disallowed = ("dag_id", "source", "source_id", "target", "target_id", "team", "team_id")
+        for key in disallowed:
+            if key in kwargs:
+                raise NotImplementedError(
+                    f"QuerySet.update() is disabled for fields ({disallowed}) to ensure cycle detection. "
+                    "Use individual save() calls instead."
+                )
+        return super().update(**kwargs)
+
+    def bulk_create(self, objs, *args, **kwargs):
+        del objs, args, kwargs  # unused
+        raise NotImplementedError("bulk_create() is disabled for Edge objects to ensure cycle detection.")
+
+    def bulk_update(self, objs, fields, *args, **kwargs):
+        disallowed = ("dag_id", "source", "source_id", "target", "target_id", "team", "team_id")
```

Comment:
This list is defined twice, here and `def update` - maybe it'd be good to pull this out to the class level so we're only maintaining one copy